### ISSUE TITLE
Error #26 (unsupported certificate purpose) in FreeRADIUS 

### ISSUE THREAD

**Issue by flipcc** [openssl/openssl#5133]
(1) prerequisites:
-OpenSSL 1.1.0g  2 Nov 2017
-FreeRADIUS Version 3.0.15
-Debian OS

(2) error:
-"SSL says error 26 : unsupported certificate purpose" in FreeRADIUS from ../ssl/statem/statem_srvr.c in line 2896 OpenSSL
-the error is thrown, when the certificate of (3) is read - when the error occurs no client certificate is known to the FreeRADIUS (and therefore, i believe, not of any interest here)

(3) details:
3.1. certficate of interest:
-Version: 3 (0x2)
-Serial Number: 4105 (0x1009)
-Signature Algorithm: sha512WithRSAEncryption
-Issuer: C = DE, ST = Sachsen, L = Leipzig, O = DBFZ Deutsches Biomasseforschungszentrum gemeinnuetzige GmbH, OU = IT, CN = DBFZ CA INTERN ROOT, emailAddress = support@dbfz.de
-Validity
--Not Before: Jan 16 12:57:19 2018 GMT
--Not After : Jan 11 12:57:19 2038 GMT
-Subject: C = DE, ST = Sachsen, L = Leipzig, O = DBFZ Deutsches Biomasseforschungszentrum gGmbH, OU = IT, CN = CAPF-91d43ef6
-Subject Public Key Info:
--Public Key Algorithm: rsaEncryption
--Public-Key: (2048 bit)
--Modulus: 00:ad:[...]
--Exponent: 65537 (0x10001)
--X509v3 extensions:
---X509v3 Subject Key Identifier: 58:A4:EB:D9:DD:CE:A2:99:72:3B:E1:20:19:1D:40:C1:F9:D5:C2:28
---X509v3 Authority Key Identifier: keyid:E2:E9:20:42:29:83:C4:77:8C:87:AB:FA:4B:A1:A9:C4:CE:00:BD:39
---X509v3 Basic Constraints: 
----critical
----CA:TRUE, pathlen:0
---X509v3 Key Usage:
----critical
----Digital Signature, Certificate Sign, CRL Sign
---X509v3 Extended Key Usage:
----critical
----TLS Web Server Authentication
---Signature Algorithm: sha512WithRSAEncryption 08:b4:[...]

3.2. end entities
-the environment includes Cisco IP phones
-the call manager demands in its (unfortunately automatically generated) CSR to be an intermediate CA with all the above keyUsages and extended keyUsages

(4) proposal:
-Cisco's approach to include that extended keyUsage is not chosen nicely, but it is out of our administrativ hands to change this
-RFC 5280 says: "In general, this [extended keyUsage] extension will appear only in end entity certificates."
-this somehow doesn't forbid Cisco to do, what they do
-therfore it should not cause an error within the OpenSSL routine that FreeRADIUS uses

**Comment by richsalz** [comment]
As was mentioned in the email discussion on this, a workaround is possbile:  write your own callback. Start with the openssl code and remove the problematic check.

**Comment by mattcaswell** [comment]
It's not clear to me from the above description what the issue actually is. Please can you expand?

**Comment by flipcc** [comment]
The actual problem is the combination of CA:true and the key usages pressed into one certificate by Cisco. 
One can help themself by expanding the enterprise PKI from windows CAs to a intermediate CA which is based on Debian and OpenSSL and allows better csr/request handling than the windows service. You would than execute:

```sh
$ openssl ca -config config/openssl.cnf -in csr/telekom/CAPF.csr -out newcerts/_testCAPF/CAPF.crt -extensions v3_intermediate_ca
```
where `openssl.cnf` has a section much like the following:

```sh
[ v3_intermediate_ca  ]
subjectKeyIdentifier = hash
authorityKeyIdentifier = keyid:always,issuer
basicConstraints = critical,CA:true,pathlen:0
keyUsage = critical,cRLSign,keyCertSign
```
this throws the problematic extension out and et voilá you wont even have problems with the Cisco system, because the certificate doesn't not need the _TLS Web Server Authentication_ usage.

**Comment by mattcaswell** [comment]
Not a 1.1.1 specific issue so removing that milestone.

**Comment by t8m** [comment]
I do not think it is a bug in OpenSSL that it rejects intermediate CA certs with the

 ---X509v3 Extended Key Usage:
----critical
----TLS Web Server Authentication

**Comment by nhorman** [comment]
Marking as inactive to be closed at the end of 3.4 dev barring further input
